json.go

```diff
@@ -0,0 +1,687 @@
+                    if (obj->fields[i].value)
+                        godror_dpiJsonNodeFree(&obj->fields[i]);
+                }
+                free(obj->fields);
+                obj->fields = NULL;
+            }
+            if (obj->fieldNames) {
+                for (i = 0; i < obj->numFields; i++) {
+                    if (obj->fieldNames[i]) {
+                        free(obj->fieldNames[i]);
+                        obj->fieldNames[i] = NULL;
+                    }
+                }
+                free(obj->fieldNames);
+                obj->fieldNames = NULL;
+            }
+            if (obj->fieldNameLengths) {
+                free(obj->fieldNameLengths);
+                obj->fieldNameLengths = NULL;
+            }
+            if (obj->fieldValues) {
+                free(obj->fieldValues);
+                obj->fieldValues = NULL;
+            }
+            break;
+    }
+}
+
+void godror_dpiJsonfreeMem(dpiJsonNode *node) {
+    if (node == NULL)
+    {
+        return;
+    }
+    godror_dpiJsonNodeFree(node);
+    godror_free_dpiNode(node);
+}
+
+*/
+import "C"
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+	"unsafe"
+)
+
+type JSONOption uint8
+
+var ErrInvalidJSON = errors.New("Invalid JSON Document")
+var ErrInvalidType = errors.New("Invalid JSON Scalar Type")
+
+const (
+	JSONOptDefault        = JSONOption(C.DPI_JSON_OPT_DEFAULT)
+	JSONOptNumberAsString = JSONOption(C.DPI_JSON_OPT_NUMBER_AS_STRING)
+	JSONOptDateAsDouble   = JSONOption(C.DPI_JSON_OPT_DATE_AS_DOUBLE)
+)
+
+// JSON holds the JSON data to/from Oracle.
+// It is like a root node in JSON tree.
+type JSON struct {
+	dpiJson *C.dpiJson
+}
+
+func (j JSON) Get(data *Data, opts JSONOption) error {
+	var node *C.dpiJsonNode
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return ErrInvalidJSON
+	}
+	jsonNodeToData(data, node)
+	return nil
+}
+
+// Returns JSONObject from JSON
+func (j JSON) GetJSONObject(opts JSONOption) (*JSONObject, error) {
+	var node *C.dpiJsonNode
+	var d Data
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return nil, ErrInvalidJSON
+	}
+	jsonNodeToData(&d, node)
+	if C.dpiOracleTypeNum(node.oracleTypeNum) != C.DPI_ORACLE_TYPE_JSON_OBJECT {
+		return nil, ErrInvalidType
+	}
+	return &JSONObject{dpiJsonObject: C.dpiData_getJsonObject(&(d.dpiData))}, nil
+}
+
+// Returns JSONArray from JSON
+func (j JSON) GetJSONArray(opts JSONOption) (*JSONArray, error) {
+	var node *C.dpiJsonNode
+	if C.dpiJson_getValue(j.dpiJson, C.uint32_t(opts), (**C.dpiJsonNode)(unsafe.Pointer(&node))) == C.DPI_FAILURE {
+		return nil, ErrInvalidJSON
+	}
+	var d Data
+	jsonNodeToData(&d, node)
+	if C.dpiOracleTypeNum(node.oracleTypeNum) != C.DPI_ORACLE_TYPE_JSON_ARRAY {
+		return nil, ErrInvalidType
+	}
+	return &JSONArray{dpiJsonArray: C.dpiData_getJsonArray(&(d.dpiData))}, nil
```

Comment:
JSONArray does not need to be a pointer (it has no pointer methods).
Maybe a JSONArray{...} would be better?

Reply:
changed it now. Yes I think we may not add more members to JSONArray which otherwise would cause member copies if pointer receiver was not used. For now, I dont see to add more members other than the dpiJsonArray pointer.